docs/internal/versioning.md

```diff
@@ -0,0 +1,244 @@
+
+```verona
+class A
+{
+	// This is always public
+	isSomething() : Bool;
+	// This is public in any API version >= 2.0
+	[[version=2.0+]]
+	isSomethingElse() : Bool;
+}
+```
+
+If a file in another package imports this then it may do something like:
+
+```verona
+import "A@1.0" as A;
+
+interface IsSomethingElse
+{
+  isSomethingElse() : Bool;
+};
+
+var a = A.A();
+```
+
+The variable `a` here does not implement the interface `IsSomethingElse` because the `isSomethingElse` method is not visible here.
+If `a` is passed to another package containing a similar interface, should it match?
+There are two obvious answers:
+
+ - If the other package imports `A` with a version <2.0, then it should not see the method.
+ - If the other package imports `A` with a version >=2.0, then it should see the method.
+
+The third case is the most interesting.
+What is the correct behaviour if the other package does not import `A` at all?
+Note that, without full dataflow analysis, it is not possible to tell in the general case which package the object was allocated in (for example, it might be stored in an array of `Any` allocated by a fourth module and populated by many other things in the system) and so the question reduces to selector identity.
+Specifically: Is the selector `isSomethingElse/1` in a module that does not import `A` the same as the selector `isSomethingElse@2.0/1` in `A`?
+There are at least two possible answers here:
+
+ - If a module does not directly import another, it sees all versioned symbols as if they were public.
+ - If a module does not directly import another, it does not see any versioned symbols.
+
+The second option is almost certainly too limiting because it effectively prevents any reliance on structural typing across packages if versioning is used.
+The first option potentially introduces breaks but is likely to avoid most of the common causes of breakage because the code in the module that is receiving objects via interface types and which cannot explicitly name the concrete types that they represent must be robust in the presence of new concrete types.
+This is more obvious when you consider the transitive case:
+
+```verona
+// In Module A:
+class A
+{
+  ...
+  [[version=1.1+]]
+  foo(Self & mut)
+};
+
+// In module B:
+interface Fooable
+{
+  foo(Self & mut);
+}
+
+interface Any {}
+
+getFooable(a : Any & mut) : Fooable | None
+{
+  match (a) {
+    Fooable => a;
+    _ => None
+  }
+}
+
+// In module C
+import "A@1.0" as A;
+import "B@2.0@ as B;
+
+// Does this return `None` or a `Fooable`?
+var a = B.getFooable(A.A());
+```
+
+If `a` at the end of this example is a `Fooable` then this is probably fine.
+The surrounding code knows that `foo` is a method on an interface that can be returned from this function, even if it didn't get one with the previous version of package `A`.
+
+The slightly more confusing outcome occurs if module `B` imports `A` with version 1.0, but `C` is updated to import version 1.1.
```

Comment:
What would be a valid use case for a dynamic update on the version of the same package on the same run time?

I was under the impression that library updates would only occur at code update, which would be replaced (together with data migration shims) and so won't see each other.

Reply:
This is about API versions, not library versions.  Imagine the case where you have a package `A` that you're using both directly and a package `B` that uses `A` indirectly.  If you update your code to use the new API, that doesn't necessarily mean that package `B` is also upgraded.  Conversely, I may pick up a new version of `B` that doesn't change its public interface at all, but has moved to using the new APIs exposed by `A`, while I am still using the old ones.

This is one of the key design goals: If `A` introduces a new API, then you don't need all of the things that depend on `A` to move to the new API at the same time.

Reply:
I see, so we also need to define what happens (and how to warn users) in cases where the user doesn't specify anything but versions break.

In your example, when `B` updates `A`, if my `A` doesn't have a version number, it could be confusing to users to be warned that their code breaks the API without modification.

```
Before:
  MyCode[0.1] -> A[any]
  B[1.0] -> A[any]
  A[1.0]

After:
  MyCode[0.1] -> A[any]
  B[2.0] -> A[2.0+]
  A[2.0] // Because "any"
```
If `MyCode` doesn't work with the new major version, and I didn't know that `B` used `A` at all, I'll get confusing error messages.

The error message I'm expecting is something like: `MyCode` doesn't work with the new version of `A (2.0)` installed as a requirement from `B(2.0)`. Please specify the version on `MyCode` to restrict it to the previous version, ie `1.0`.

Instead, if not clear, the error message could be something horrible like: `A::foobar` not found, from `MyCode::functionX()`. Even though the only thing you did was "tool update B".

But this requires a knowledge of the previous environment and the reasons why all versions are what they are, which isn't trivial to communicate effectively to users.

Also, expecting users to manage that, even when using tools to dump the state, on large projects, can be daunting.

Reply:
If we instead, say: any module that upgrades another, bumps the local version but leaves other versions intact.

For example:
```
Before:
  MyCode[0.1] -> A[any(weak: 1.0)]
  B[1.0] -> A[any(weak: 1.0)]
  A[(weak: 1.0)]

After:
  MyCode[0.1] -> A[any(strong: 1.0)]
  B[2.0] -> A[forced:2.0+)]
  A[(strong:1.0) + (forced:2.0)] // Because local upgrade
```

This will never break, but can lead to a huge number of variants without actual need.

Reply:
I am not sure what the numbers refer to in your snippet.  Are these the implementation versions or API versions?  If they are the API versions, which package's API versions are they referring to?

Reply:
Sorry, it was a bit confusing. The numbers are API versions with the following rules:
* Just numbers (ex. `MyCode[1.0]`) means a fixed version, by the user
* An arrow means "needs", so `MyCode[1.0] -> A[...]` means `MyCode` needs some `A`
* `any` means I don't care which version, any should work (or so my ignorance says so)
* `weak` means the user didn't request specific versions and the compiler inferred something
* `strong` means the compiler was forced to choose a specific version due to constraints
* `forced` means the user forced the compiler to choose a specific version

So:
```
// MyCode version 0.1 just needs A, doesn't care what, compiler picks 1.0 because it's what's available
MyCode[0.1] -> A[any(weak: 1.0)]

// B version 1.0 needs A in the same way, same choice
B[1.0] -> A[any(weak: 1.0)]

// A is imported in version "weak" 1.0 (ie, no constraints)
A[(weak: 1.0)]

// =========== We migrate B to version 2.0, requiring A version 2.0+

// B was upgraded to 2.0 and it now requires A version 2.0 or greater
// We must import that, too, but there could be others (MyCode) that still use 1.0
B[2.0] -> A[forced:2.0+)]

// MyCode version 1.0 still depends on A, but we don't know if it works with 2.0
// and the user didn't say anything so we add a strong requirement to force the 1.0
// Unless the policy is to upgrade and let users handle subtle breakages on their own?
MyCode[0.1] -> A[any(strong: 1.0)]

// Now we have two versions of A
A[(strong:1.0) + (forced:2.0)] // From B and MyCode
```

Reply:
The proposal for API versions is that you must *always* specify an API version when you import a package (including the `Builtin` package).  The problems happen if one package imports another (with a version), and then passes pointers to concrete types from that package as interface types to another.

Reply:
I see.

It can be cumbersome to specify all versions when you don't care about it, or there's only ever going to be one (I control it). So I imagine that, in theory, the helper tool that creates the config file would do that for you, and you just check that everything is right.

Then API version migration would generate a different config file that would show up in a `git diff`, making it easy for developers to know what changed and take action if needed.